doc/sphinx/techref/pyextend.rst

```diff
@@ -0,0 +1,438 @@
+:py:meth:`fontforge.scriptPath() <fontforge.scriptPath>` function, is
+automatically run when FontForge is started.
+
+As a simple example, here is a script that adds an "Add Midpoint Contour" entry
+to the Char View Tools menu. When any two on-curve points of a glyph are
+selected this adds a new contour containing a single point that is located
+midway between the two selected points.::
+
+    import fontforge
+
+    def getSelectedPoints(l):
+        pl = []
+        for c in l:
+            pl.extend( [ p for p in c if p.on_curve and p.selected ] )
+        return pl
+
+    def midContourEnable(u, glyph):
+        return len(getSelectedPoints(glyph.layers[glyph.activeLayer]))==2
+
+    def addMidContour(u, glyph):
+        layer_id = glyph.activeLayer
+        l = glyph.layers[layer_id]
+        pl = getSelectedPoints(l)
+        if len(pl) != 2:
+            fontforge.postError("Bad selection", "You must select "
+            "exactly two on-curve points to add a midpoint contour.")
+            return
+        nc = fontforge.contour()
+        nc.insertPoint(((pl[0].x+pl[1].x)/2, (pl[0].y+pl[1].y)/2, True,
+                fontforge.splineCorner, True))
+        l += nc
+        glyph.preserveLayerAsUndo(layer_id)
+        glyph.layers[layer_id] = l
+
+    fontforge.registerMenuItem(addMidContour, midContourEnable,
+            None, ("Glyph"), None, "Add Midpoint Contour")
+
+This script is typical in that it starts with some imports, defines some
+functions, and ends by invoking :py:meth:`fontforge.registerMenuItem`. Other
+scripts may import other modules, define more elaborate functions or classes,
+or register more menu items or other callbacks. Still, this is a common
+general pattern.
+
+The first step in converting the script to a plugin is to wrap the last
+section in a function called ``fontforge_plugin_init()``. Once the file
+is packaged appropriately this function will be called after FontForge
+discovers and loads the plugin (if the user has Enabled it).::
+
+    def fontforge_plugin_init(**kw):
+        fontforge.registerMenuItem(addMidContour, midContourEnable,
+                None, ("Glyph"), None, "Add Midpoint Contour")
+
+The ``**kw`` function argument will capture any keyword arguments passed
+to the function. There is currently one, discussed below, but more may
+be added in the future. (Note that while the function can be located
+in other places (see the section on Entry-points below) it must be a
+*function*, not a *method*.
+
+The remaining steps have to do with python "packaging", which is too
+large a topic to discuss extensively in this guide. A good starting
+point is `Packaging Python Projects
+<https://packaging.python.org/tutorials/packaging-projects/>`_ from the
+Python documentation. The following are the very basics.
+
+First we create a new directory for our project with the following files:
+
+1. ``fontforge_midpoint.py`` (The script itself as a single module file.)
```

Comment:
One thing you can do is to create something like a [cookiecutter](https://github.com/cookiecutter/cookiecutter) template that allows you to quickly bootstrap a python package. It's what I use at work, and it's pretty handy in saving you from doing a lot of the boilerplate.